server-openid4vci/src/main/java/com/android/identity/server/openid4vci/AuthorizeServlet.kt

```diff
@@ -16,56 +40,147 @@ import kotlin.time.Duration.Companion.minutes
         val requestUri = req.getParameter("request_uri")
         val code = requestUri.substring(requestUri.lastIndexOf(":") + 1)
         val id = codeToId(OpaqueIdType.PAR_CODE, code)
-        val stateStr = idToCode(OpaqueIdType.AUTHORIZATION_STATE, id, 20.minutes)
+        val authorizationCode = idToCode(OpaqueIdType.AUTHORIZATION_STATE, id, 20.minutes)
+        val pidReadingCode = idToCode(OpaqueIdType.PID_READING, id, 20.minutes)
+        val authorizeHtml = resources.getStringResource("$RESOURCE_BASE/authorize.html")!!
         resp.contentType = "text/html"
-        resp.writer.print("""
-            <!DOCTYPE html>
-            <html>
-            <head>
-            <meta name="viewport" content="width=device-width, initial-scale=1.0" />
-            </head>
-            <body>
-                <h1>Enter your name</h1>
-                <form action="authorize" method="post">
-                  <label for="f">First name:</label>
-                  <input type="text" id="f" name="f"><br/>
-                  <label for="l">Last name:</label>
-                  <input type="text" id="l" name="l"><br/>
-                  <input type="hidden" name="state" value="$stateStr"/>
-                  <input type="submit" value="Submit">
-                </form>
-                </form>
-            </body>
-            </html>
-        """.trimIndent())
+        resp.writer.print(
+            authorizeHtml
+                .replace("\$authorizationCode", authorizationCode)
+                .replace("\$pidReadingCode", pidReadingCode))
     }
 
     /**
      * Handle first/last name form submission and redirect to [FinishAuthorizationServlet].
      */
     override fun doPost(req: HttpServletRequest, resp: HttpServletResponse) {
-        val code = req.getParameter("state")
+        val code = req.getParameter("authorizationCode")
+        val pidData = req.getParameter("pidData")
+        val extraInfo = req.getParameter("extraInfo")
         val id = codeToId(OpaqueIdType.AUTHORIZATION_STATE, code)
         val storage = environment.getInterface(Storage::class)!!
+        val configuration = environment.getInterface(Configuration::class)!!
+
+        val tokenData = Json.parseToJsonElement(pidData).jsonObject["token"]!!
+            .jsonPrimitive.content.fromBase64Url()
+
+        val (cipherText, encapsulatedPublicKey) = parseCredentialDocument(tokenData)
+
         runBlocking {
+            val baseUri = URI(configuration.getValue("base_url")!!)
+            val origin = baseUri.scheme + "://" + baseUri.authority
             val state = IssuanceState.fromCbor(storage.get("IssuanceState", "", id)!!.toByteArray())
-            state.credentialData = CredentialData(
-                firstName = req.getParameter("f"),
-                lastName = req.getParameter("l")
+            val uncompressed = (state.pidReadingKey!!.publicKey as EcPublicKeyDoubleCoordinate).asUncompressedPointEncoding
```

Comment:
nit: Can we give this a clearer name? uncompressedKeyPoint? ...or similar?

Reply:
Done